hail/python/hail/ggplot/ggplot.py

```diff
@@ -172,30 +172,27 @@ def get_aggregation_result(selected, mapping_per_geom, precomputed):
         fig = go.Figure()
 
         for geom, (geom_label, agg_result) in zip(self.geoms, aggregated.items()):
-            listified_agg_result = labels_to_stats[geom_label].listify(agg_result)
+            df_agg_result = labels_to_stats[geom_label].listify(agg_result)
 
-            if listified_agg_result:
-                relevant_aesthetics = [scale_name for scale_name in list(listified_agg_result[0]) if scale_name in self.scales]
+            if not df_agg_result.empty:
+                relevant_aesthetics = [scale_name for scale_name in df_agg_result.columns if scale_name in self.scales]
                 for relevant_aesthetic in relevant_aesthetics:
-                    listified_agg_result = self.scales[relevant_aesthetic].transform_data_local(listified_agg_result, self)
+                    df_agg_result = self.scales[relevant_aesthetic].transform_data_local(df_agg_result, self)
                 # Need to identify every possible combination of discrete scale values.
-                discrete_aesthetics = [scale_name for scale_name in relevant_aesthetics if self.scales[scale_name].is_discrete() and scale_name != "x"]
-                subsetted_to_discrete = tuple([one_struct.select(*discrete_aesthetics) for one_struct in listified_agg_result])
+                grouping_aesthetics = [scale_name for scale_name in relevant_aesthetics if should_use_scale_for_grouping(self.scales[scale_name]) and scale_name != "x"]
+
+                subsetted_to_discrete = df_agg_result[grouping_aesthetics]
+
                 group_counter = 0
 
                 def increment_and_return_old():
                     nonlocal group_counter
                     group_counter += 1
                     return group_counter - 1
                 numberer = defaultdict(increment_and_return_old)
-                numbering = [numberer[data_tuple] for data_tuple in subsetted_to_discrete]
-
-                numbered_result = []
-                for s, i in zip(listified_agg_result, numbering):
-                    numbered_result.append(s.annotate(group=i))
-                listified_agg_result = numbered_result
+                df_agg_result["group"] = [numberer[tuple(x)] for _, x in subsetted_to_discrete.iterrows()]
```

Comment:
You can leave this for future improvement, but I suspect you could do this more easily and faster with pandas `groupby`.

Reply:
I am going to do this in future improvement, since I'm currently in the midst of refactoring this a bit due to a bug report about running out of colors to plot with.